pinot-core/src/main/java/com/linkedin/pinot/core/util/DataTableCustomSerDe.java

```diff
@@ -0,0 +1,417 @@
+   *
+   * @param doubleArray DoubleArray to serialize
+   * @return Serialized value for the doubleArray
+   */
+  public static byte[] serializeDoubleArrayList(DoubleArrayList doubleArray) {
+    int numElements = doubleArray.size();
+    int size = (V1Constants.Numbers.INTEGER_SIZE) + (numElements * (V1Constants.Numbers.DOUBLE_SIZE));
+
+    ByteBuffer byteBuffer = ByteBuffer.allocate(size);
+    byteBuffer.putInt(numElements);
+
+    double[] elements = doubleArray.elements();
+    for (int i = 0; i < numElements; i++) {
+      byteBuffer.putDouble(elements[i]);
+    }
+    return byteBuffer.array();
+  }
+
+  /**
+   * Helper method to de-serialize DoubleArrayList
+   *
+   * @param bytes Serialized bytes for DoubleArrayList
+   * @return De-serialized DoubleArrayList from byte-array.
+   */
+  public static DoubleArrayList deserializeDoubleArray(byte[] bytes) {
+    ByteBuffer byteBuffer = ByteBuffer.wrap(bytes);
+
+    int numElements = byteBuffer.getInt();
+    DoubleArrayList doubleArray = new DoubleArrayList(numElements);
+
+    for (int i = 0; i < numElements; i++) {
+      doubleArray.add(byteBuffer.getDouble());
+    }
+    return doubleArray;
+  }
+
+  /**
+   * Helper method to serialize a HashMap into a byte-array.
+   *
+   * @param map Map to serialize
+   * @return Serialized byte-array for the HashMap.
+   */
+  public static byte[] serializeHashMap(HashMap<Object, Object> map) {
+    ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+    DataOutputStream dataOutputStream = new DataOutputStream(byteArrayOutputStream);
+
+    try {
+      // Write the size of the map.
+      dataOutputStream.writeInt(map.size());
+
+      // Write the serialized key-value pairs
+      boolean first = true;
+      for (Map.Entry<Object, Object> entry : map.entrySet()) {
+
+        // Write the key and value type before writing the first key-value pair.
+        if (first) {
+          dataOutputStream.writeInt(getDataTypeOfObject(entry.getKey()).getValue());
+          dataOutputStream.writeInt(getDataTypeOfObject(entry.getValue()).getValue());
+          first = false;
+        }
+
+        byte[] bytes = serializeObject(entry.getKey());
+        dataOutputStream.writeInt(bytes.length);
+        dataOutputStream.write(bytes);
+
+        bytes = serializeObject(entry.getValue());
+        dataOutputStream.writeInt(bytes.length);
+        dataOutputStream.write(bytes);
+      }
+
+      dataOutputStream.flush();
+    } catch (IOException e) {
+      LOGGER.error("Exception caught while serializing HashMap", e);
+      throw new RuntimeException("HashMap serialization error");
+    }
+
+    return byteArrayOutputStream.toByteArray();
+  }
+
+  /**
+   * Helper method ro de-serialize a HashMap
+   *
+   * @param bytes Serialized bytes for the HashMap
+   * @return De-serialized HashMap from the byte-array
+   */
+  public static HashMap deserializeHashMap(byte[] bytes) {
+    ByteBuffer byteBuffer = ByteBuffer.wrap(bytes);
```

Comment:
Where is the check/guarantee provided ? The deserialize method that calls into this does not check that. 

Preconditions or if checks are problem if those are executed in tight loop for many rows (almost any operation is a perf hit there). One extra operation here will have negligible impact